### DISCUSSION TITLE
Is there a way to display a Godot scene 'in' the UI? (Basically Godot's SubViewport Control)

### DISCUSSION THREAD

**Discussion by MechWarrior99** [MrJul/Estragonia#7] [Q&A] [answered]
Hello, I got a pretty simple question, hopefully with a simple answer. I was wondering if there is a way to display a scene 'in' the Avalonia UI as apposed to just overlaying the UI on a scene? Like the SubViewport control in Godot.

Thanks! :)

**Comment by MrJul** [comment]
Hello,

Unfortunately, this isn't supported: there is currently no way to embed Godot in Avalonia. This is basically the reverse of Estragonia, which I'm interested in developing once we have a proper [Godot library](https://github.com/godotengine/godot/pull/72883).

Note that Estragonia uses a Godot SubViewport to render the Avalonia's tree, so you can reuse its texture where you want in Godot (for example have your UI displayed on the mesh of your choice), it's not only a simple screen overlay.

**Reply by MechWarrior99** [reply]
Hey, thanks for the reply.

Yeah that makes sense, and basically what I was figuring. I wonder how or if it would be possible to render a Godot viewport to a texture and then take that and feed it to a Avalonia control to render, and without killing performance...

**Reply by MrJul** [reply, marked as answer]
It should be possible in theory. They're all Vulkan images at the end of the day, so one can render a Godot viewport to a texture, get the underlying Vulkan image, and use Avalonia's GPU interop to manipulate it. Setting up the correct Vulkan fences to get all that to render correctly is probably the hardest part.

What use case do you have in mind? You should be able to render Avalonia in Godot using Estragonia today, add some control as a placeholder to where you need a scene to be rendered, and transform a Godot's SubViewport to match its position and size - all that without any new GPU interop. A bit convoluted, but it should work.

**Reply by MechWarrior99** [reply]
Yeah that sounds rough. The idea of keeping a viewport position in sync with a control position was the other way I had thought of as well, but seemed like (at least on initial impression) to be a bit less 'stable' option. But that does seem like a easier solution, if a bit messy.

I'm working on a standalone application for procedural mesh generation. So trying to get a 'scene view' for rendering the mesh.
At this point, I am really only using Godot for 3D rendering and building. If there was a good high level C# package (or wrapper) for 3D rendering (I'm not very familiar with graphics APIs) I would probably just switch from Godot. But right now, this seems like the best option. Anyways, that is neither here nor there.

**Comment by GeorgeS2019** [comment]
https://github.com/AvaloniaUI/Avalonia/discussions/12709#discussioncomment-6884828
